src/esrc.rs

```diff
@@ -0,0 +1,195 @@
+    let x1 = Scalar::from(rng.next_u64()).double();
+    let x_hat0 = G2Affine::generator() * x0;
+    let x_hat1 = G2Affine::generator() * x1;
+    (SigningKey(x0, x1), VerifyKey(x_hat0, x_hat1))
+}
+
+#[allow(dead_code)]
+fn encrypt(
+    rng: &mut (impl RngCore + CryptoRng),
+    m: G1Affine,
+    enc_key: &EncryptionKey,
+) -> CipherText {
+    let r = Scalar::from(rng.next_u64());
+    let c0 = G1Affine::generator() * r;
+    let c1 = m + (enc_key * r);
+    CipherText(c0, c1)
+}
+
+#[allow(dead_code)]
+fn decrypt(cm: &CipherText, dec_key: &DecryptionKey) -> G1Affine {
+    let CipherText(c0, c1) = cm;
+    let dec = c1 - (c0 * dec_key);
+    G1Affine::from(dec)
+}
+
+#[allow(dead_code)]
+fn randomize(cm: &CipherText, enc_key: &EncryptionKey, r_prime: Scalar) -> CipherText {
+    let CipherText(c0, c1) = cm;
+    let rc0 = c0 + (G1Affine::generator() * r_prime);
+    let rc1 = c1 + (enc_key * r_prime);
+    CipherText(rc0, rc1)
+}
+
+#[allow(dead_code)]
+fn sign(
+    rng: &mut (impl RngCore + CryptoRng),
+    cm: &CipherText,
+    enc_key: &EncryptionKey,
+    sign_key: SigningKey,
+) -> Signature {
+    let scalar_s = Scalar::from(rng.next_u64());
+    let scalar_s_inverted = scalar_s.invert().unwrap(); // TODO: how to handle?
+    let CipherText(c0, c1) = cm;
+    let SigningKey(x0, x1) = sign_key;
+    let z = (G1Affine::generator() + (c0 * x0) + (c1 * x1)) * scalar_s_inverted;
+    let s = G1Affine::generator() * scalar_s;
+    let s_hat = G2Affine::generator() * scalar_s;
+    let t = ((G1Affine::generator() * x0) + (enc_key * x1)) * scalar_s_inverted;
+    Signature(z, s, s_hat, t)
+}
+
+#[allow(dead_code)]
+fn adapt(
+    rng: &mut (impl RngCore + CryptoRng),
+    signature: &Signature,
+    r_prime: Scalar,
+) -> Signature {
+    let scalar_s_prime = Scalar::from(rng.next_u64());
+    let scalar_s_prime_inverted = scalar_s_prime.invert().unwrap(); // TODO: how to handle?
+    let Signature(z, s, s_hat, t) = signature;
+
+    let z_prime = (z + (t * r_prime)) * scalar_s_prime_inverted;
+    let s_prime = s * scalar_s_prime;
+    let s_hat_prime = s_hat * scalar_s_prime;
+    let t_prime = t * scalar_s_prime_inverted;
+    Signature(z_prime, s_prime, s_hat_prime, t_prime)
+}
+
+#[allow(dead_code, clippy::similar_names)] // g_g_hat_pair and g_s_hat_pair
+fn verify(
+    verify_key: &VerifyKey,
+    enc_key: &EncryptionKey,
+    cm: &CipherText,
+    signature: &Signature,
+) -> bool {
+    let Signature(z, s, s_hat, t) = signature;
+    let CipherText(c0, c1) = cm;
+    let VerifyKey(x_hat0, x_hat1) = verify_key;
+
+    let z_s_pair = pairing(&G1Affine::from(z), &G2Affine::from(s_hat));
+    let g_g_hat_pair = pairing(&G1Affine::generator(), &G2Affine::generator());
+    let c0_x_hat0_pair = pairing(&G1Affine::from(c0), &G2Affine::from(x_hat0));
+    let c1_x_hat1_pair = pairing(&G1Affine::from(c1), &G2Affine::from(x_hat1));
+    if z_s_pair != g_g_hat_pair + c0_x_hat0_pair + c1_x_hat1_pair {
```

Comment:
Is addition the right operator for pairings? (Just double checking. The paper just has them next to each other, so I'd assume that's multiplication, but it might just be the default operator, and that could be addition in this case.)